mm2src/coins/rpc_command/init_scan_for_new_addresses.rs

```diff
@@ -0,0 +1,152 @@
+pub type ScanAddressesTaskHandle = RpcTaskHandle<InitScanAddressesTask>;
+pub type ScanAddressesRpcTaskStatus = RpcTaskStatus<
+    ScanAddressesResponse,
+    HDAccountBalanceRpcError,
+    ScanAddressesInProgressStatus,
+    ScanAddressesAwaitingStatus,
+>;
+
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct ScanAddressesResponse {
+    pub account_index: u32,
+    pub derivation_path: RpcDerivationPath,
+    pub new_addresses: Vec<HDAddressBalance>,
+}
+
+#[derive(Deserialize)]
+pub struct ScanAddressesRequest {
+    coin: String,
+    #[serde(flatten)]
+    params: ScanAddressesParams,
+}
+
+#[derive(Deserialize)]
+pub struct ScanAddressesParams {
+    pub account_index: u32,
+    pub gap_limit: Option<u32>,
+}
+
+#[derive(Clone, Serialize)]
+pub enum ScanAddressesInProgressStatus {
+    InProgress,
+}
+
+/// We can't use `std::convert::Infallible` as [`RpcTaskTypes::UserAction`] because it doesn't implement `Serialize`.
+/// Use `!` when it's stable.
+#[derive(Clone, Serialize)]
+pub enum ScanAddressesUserAction {}
+
+/// We can't use `std::convert::Infallible` as [`RpcTaskTypes::AwaitingStatus`] because it doesn't implement `Serialize`.
+/// Use `!` when it's stable.
+#[derive(Clone, Serialize)]
+pub enum ScanAddressesAwaitingStatus {}
+
+#[async_trait]
+pub trait InitScanAddressesRpcOps {
+    async fn init_scan_for_new_addresses_rpc(
+        &self,
+        params: ScanAddressesParams,
+    ) -> MmResult<ScanAddressesResponse, HDAccountBalanceRpcError>;
+}
+
+pub struct InitScanAddressesTask {
+    req: ScanAddressesRequest,
+    coin: MmCoinEnum,
+}
+
+impl RpcTaskTypes for InitScanAddressesTask {
+    type Item = ScanAddressesResponse;
+    type Error = HDAccountBalanceRpcError;
+    type InProgressStatus = ScanAddressesInProgressStatus;
+    type AwaitingStatus = ScanAddressesAwaitingStatus;
+    type UserAction = ScanAddressesUserAction;
+}
+
+#[async_trait]
+impl RpcTask for InitScanAddressesTask {
+    fn initial_status(&self) -> Self::InProgressStatus { ScanAddressesInProgressStatus::InProgress }
+
+    async fn run(self, _task_handle: &ScanAddressesTaskHandle) -> Result<Self::Item, MmError<Self::Error>> {
+        match self.coin {
+            MmCoinEnum::UtxoCoin(utxo) => utxo.init_scan_for_new_addresses_rpc(self.req.params).await,
+            MmCoinEnum::QtumCoin(qtum) => qtum.init_scan_for_new_addresses_rpc(self.req.params).await,
+            _ => MmError::err(HDAccountBalanceRpcError::CoinIsActivatedNotWithHDWallet),
+        }
+    }
+}
+
+pub async fn init_scan_for_new_addresses(
+    ctx: MmArc,
+    req: ScanAddressesRequest,
+) -> MmResult<InitRpcTaskResponse, HDAccountBalanceRpcError> {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin).await?;
+    let coins_ctx = CoinsContext::from_ctx(&ctx).map_to_mm(HDAccountBalanceRpcError::Internal)?;
+    let task = InitScanAddressesTask { req, coin };
+    let task_id = ScanAddressesTaskManager::spawn_rpc_task(&coins_ctx.scan_addresses_manager, task)?;
```

Comment:
rustfmt splits this line into 4:
```rust
let task_id = ScanAddressesTaskManager::spawn_rpc_task(&coins_ctx.scan_addresses_manager, InitScanAddressesTask {
        req,
        coin,
    })?;
```

Reply:
> rustfmt splits this line into 4:
> 
> ```rust
> let task_id = ScanAddressesTaskManager::spawn_rpc_task(&coins_ctx.scan_addresses_manager, InitScanAddressesTask {
>         req,
>         coin,
>     })?;
> ```

I wouldn't take that variable cost to avoid that. Just my opinion.

Reply:
Hmmm, I'll try to figure out how much it may cost

Reply:
The following statements are valid for Release mode.
First of all, I'd like to give an example:
```rust
let x = Vec::with_capacity(10);
let y = x;
println!("{}", y);
```
As it's described at the [More Operations That Move](https://www.oreilly.com/library/view/programming-rust/9781491927274/ch04.html) section:
> the Rust compiler’s code generation is good at “seeing through” all these moves; in practice, the machine code often stores the value directly where it belongs.

And if we can trust an [answer on StackOverflow](https://stackoverflow.com/questions/30288782/what-are-move-semantics-in-rust)
>  the compiler might apply "return value optimization" and build directly in the caller's stack frame -- that is, the caller passes a pointer into which to write the return value, which is used without intermediary storage.

So I think Rust's compiler is smart enough to avoid unnecessary `memcpy` of the `InitScanAddressesTask` structure.
But even if it copies the memory of the structure
https://github.com/KomodoPlatform/atomicDEX-API/blob/7224b85e39a2406b5b9e8cb9a8e69cd59fda514f/mm2src/coins/rpc_command/init_scan_for_new_addresses.rs#L64-L67
Where `MmCoinEnum` is a wrapper over an Arc<Coin>, and `std::mem::size_of::<ScanAddressesRequest>()` equals to 40 bytes which is not so critical in my opinion.

Especially copying of integers is much more efficient than using `Iterator::collect`, for example, into [HashSet](https://doc.rust-lang.org/src/std/collections/hash/set.rs.html#997-1008)
```rust
    #[inline]
    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> HashSet<T, S> {
        let mut set = HashSet::with_hasher(Default::default());
        set.extend(iter);
        set
    }
```
We could use `for` loop instead of `Iterator::collect` when we know the exact number of items along with `HashSet::with_capacity`, but don't do it because it looks so much harder.

Reply:
In my opinion, it's better to rely on compiler optimization here:
```rust
let task = let task = InitScanAddressesTask { req, coin };
let task_id = ScanAddressesTaskManager::spawn_rpc_task(&coins_ctx.scan_addresses_manager, task)?;
```
where it's clear what task actually is spawned, and it requires 2 lines instead of:
```
let task_id = ScanAddressesTaskManager::spawn_rpc_task(&coins_ctx.scan_addresses_manager, 
    InitScanAddressesTask {
        req,
        coin,
    })?;
```

Reply:
Actually, almost every compilers are smart to deal things like that like I described above. But, my point is "If it doesn't increase the complexity too much, avoid creating unnecessary variables to increase the readability as much as possible.". So we have to make a good balance between overkilling the project with not creating variables to apply micro-optimizations to the project, and overkilling the project with creating variables just because it could increase the readibility.

The why is pretty simple. I will try my best to explain it with giving explanations, examples, and benchmark results.

First of all, even if we wouldn't care about the runtime performance(which we definetly do but just as an example), this will still cost to our project as increasing the compilation time which is currently one of the critical problem in our project([see #1269](https://github.com/KomodoPlatform/atomicDEX-API/issues/1269)). If we can gain compilation time by losing a little bit readibility(may even increase it in many cases), this could be pretty good trade for our project.

Secondly, the codebase will be little bit smaller.

Thirdly, we can't never know the behaviour of the compilers specially ones that provides high level abstractions/functionalities over syscalls. There are lots of functionalities happening behind the hood to provide that magical benefits. You can't really debug it. Because non of them works flawlessly since it's created by humans. :) It would be better to not leave complete control of the project to the compilers. I believe we are(API team) more than enough to take some optimization control from compiler and do it for ourselves. Which might be applying inline-cold attributes, decreasing memory allocations, unsafe(in rusty term, not really unsafe) implementations, etc.

I can give an example in practical way. Lets consider two of the the following rust applications.
```rs
fn main() {
    let x: std::collections::HashSet<i32> = vec![1i32, 2, 3].into_iter().collect();
    println!("{:?}", x);
}
```
assembly output(by executing `cargo rustc --release -- --emit asm` command) of the code above -> https://github.com/ozkanonur/rust-assembly-debug/blob/master/no-move.s

```rs
fn main() {
    let x: std::collections::HashSet<i32> = vec![1i32, 2, 3].into_iter().collect();
    let y = x; // value is moved here
    println!("{:?}", y);
}
```
assembly output(by executing `cargo rustc --release -- --emit asm` command) of the code above -> https://github.com/ozkanonur/rust-assembly-debug/blob/master/move.s

```diff
358,359c358,365
<       movq    %rsp, %rax
<       movq    %rax, 120(%rsp)
---
>       movaps  (%rsp), %xmm0
>       movaps  16(%rsp), %xmm1
>       movaps  32(%rsp), %xmm2
>       movaps  %xmm2, 160(%rsp)
>       movaps  %xmm1, 144(%rsp)
>       movaps  %xmm0, 128(%rsp)
>       leaq    128(%rsp), %rax
>       movq    %rax, 112(%rsp)
361c367
```

We expect the compiler to resolve this simple algorithm(keep working on directly address of x) and not do execute extra instructions on memory. But what really happens is, each value of the hashset will be moved one-by-one to the xmm register then to the rsp. Right after that beggining of the hashset value address will be moved to the rax register and then rax register will be moved to the rsp.

In total, we will have 6 extra assembly instructions in `let y = x;` example. In larger projects, we would have thousands of unnecessary instructions in our project just by relying on the compiler about this move variable thing. Which is pretty suicide for low-power(maybe arm cpus) devices.

By the way, this isn't related to `InitScanAddressesTask` codeblock, I just wanted to share experiences and honest opinions about this topic. For `InitScanAddressesTask`, maybe compiler will not do the same thing that it did in my example. But we can't really know without dissassembling it. So I guess that's an impposible task for mm2 project :)

Please feel free to keep or update it. The decision is yours. This is very minimal optimization so we can have exceptions in cases like this. Like I said, we should make a balance to keep the project beautiful and optimized.

Reply:
Thank you for the explanation! I understand your point of view and now I agree with that. Such micro optimizations can improve not only runtime performance, but also the compile time.

I'm not familiar with Assembler at all, but I guess these instructions
```asm
>       movaps  (%rsp), %xmm0
>       movaps  16(%rsp), %xmm1
>       movaps  32(%rsp), %xmm2
>       movaps  %xmm2, 160(%rsp)
>       movaps  %xmm1, 144(%rsp)
>       movaps  %xmm0, 128(%rsp)
```
move the value of the memory allocated on stack `rsp` for the `HashSet` structure (that is actually 48 bytes) in parts of 16 bytes 3 times. But items of `HashSet` are allocated on the heap, so we copy only such fields as `pointer to the heap`, `capacity`, `len` etc.
Please correct me if I'm wrong.

> Secondly, the codebase will be little bit smaller.

In some places such optimizations can make the codebase shorter, and there are a lot to optimize in this PR like 
this https://github.com/KomodoPlatform/atomicDEX-API/pull/1255#discussion_r862971404

But in this case it doesn't make it shorter 😅 
Thank you again, I'll try to avoid using unnecessary variables where they would be redundant 👍

Reply:
> move the value of the memory allocated on the stack `rsp` for the `HashSet` structure (that is actually 48 bytes) in parts of 16 bytes 3 times. But items of `HashSet` are allocated on the heap, so we copy only such fields as `pointer to the heap`, `capacity`, `len` etc. Please correct me if I'm wrong.

That's correct.

> But in this case it doesn't make it shorter sweat_smile It even makes the code difficult to read.

Yeah for this part of code true. I was also trying to give overall looking, It wasn't about this codeblock :slight_smile: